velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+and helps them manage their spilling state on disk. There is one Spiller
+object created for each operator. The Spiller object takes a row container
+object on construction (except hash probe operator which doesn’t buffer inputs
+and directly writes spilling rows to disk). The row container is a row-wise
+in-memory data store owned by the operator to store its in-memory state that
+could be spilled out to disk when there is a lack of memory. For example, the
+row container of a hash aggregation operator stores the intermediate
+aggregation state with one row per each group by key. When the operator
+requests to spill, the Spiller scans through all the rows in the row container
+and calculates each row’s partition number. Then the Spiller decides which
+partitions need to spill, and then write them out as a list of files. The
+Spiller also helps operators like hash aggregation to sort the spill data
+before writing to disk. Each spill file contains a single sorted run. On
+restore, the Spiller will help the operator to read back the spilled data to
+restore the in-memory state. It creates a merge sort reader to read the sorted
+data and an unsorted reader to read the unsorted data.
+
+Spiller implements the following major functions:
+
+**Spill Data Partition**: when spilling, we only want to spill a minimum amount
+of data to disk and keep the rest of the operator state in-memory. This
+requires the Spiller to divide the rows in the row container into a number of
+partitions and only spill some of them as needed. This does not only reduce the
+spill io cost but also speed up the restore procedure. For un-spilled
+partitions, we can process them all together with the normal fast execution
+path. For spilled partitions, each will have its own set of files on disk and
+will be restored separately. The columns used to calculate the partition number
+are specified by the operator based on the spilling algorithm. Take hash
+aggregation for example, it sets the partition columns to the group by keys.
+This ensures the rows with the same group by key will spill to the same
+partition and then be restored together.
+
+**spill partition selection**: we select partitions with the most spillable data to
+spill initially, and then stick with those partitions if they have sufficient
+memory to free up. The spillable data is measured as the memory bytes occupied
+by the rows in the row container. For operators which need sorted spilling, we
+might need to avoid spilling from partitions which have a very small amount of
+spillable data even though it has already been spilled. The reason is that the
+sorted spilling requires that each spill file contains a single sorted run
+which means we can’t continuously append data to the same spill file. There is
+no such restriction for non-sorted spilling such as for the hash build
+operator.
+
+**spill data sort**: the operator that needs sorted spilling will run sort merge
```

Comment:
> **spill data sort**

Sort data while spilling

> the operator that needs sorted spilling will run sort merge

Operators that use sorted spill combine spill files with in-memory data using sort merge algorithm.